vendor/src/github.com/docker/libnetwork/drivers/routed/netfilter.go

```diff
@@ -0,0 +1,178 @@
+		}
+	}
+	return nil
+}
+
+func (r *IPRange) String() string {
+	return r.from.String() + "-" + r.to.String()
+}
+
+type netFilterConfig struct {
+	allowedNets   []*net.IPNet
+	allowedRanges []*IPRange
+}
+
+type netFilter struct {
+	ifaceName string
+	config    *netFilterConfig
+}
+
+func ParseIpOrNet(ipStr string) *net.IPNet {
+	if !strings.Contains(ipStr, "/") {
+		ipStr += "/32"
+	}
+
+	if _, ipNet, err := net.ParseCIDR(ipStr); err == nil {
+		return ipNet
+	} else {
+		return nil
+	}
+}
+
+func NetFilterConfigParse(ingressAllowedString string) (*netFilterConfig, error) {
+	config := new(netFilterConfig)
+	if ingressAllowedString != "" {
+		for _, filterElement := range strings.Split(ingressAllowedString, ",") {
+			filterElement = strings.TrimSpace(filterElement)
+			ipNet := ParseIpOrNet(filterElement)
+			if ipNet == nil {
+				if ipRange := ParseIPRange(filterElement); ipRange != nil {
+					config.allowedRanges = append(config.allowedRanges, ipRange)
+				} else {
+					return nil, fmt.Errorf("NetFilter: Could not parse IP, CIDR or IPRange %s", filterElement)
+				}
+			} else {
+				config.allowedNets = append(config.allowedNets, ipNet)
+			}
+		}
+	}
+	return config, nil
+}
+
+func NewNetFilter(ifaceName string, epOptions map[string]interface{}) *netFilter {
+	logrus.Debugf("New NetFilter for iface %s and options %s", ifaceName, epOptions)
+
+	ingressFiltering := epOptions[netlabel.IngressAllowed].(*netFilterConfig)
+	if ingressFiltering == nil {
+		logrus.Info("No network ingress filtering specified")
+	}
+
+	return &netFilter{ifaceName, ingressFiltering}
+}
+
+func chainExists(chainName string) bool {
+	return iptables.Exists("", chainName, "-N", chainName)
+}
+
+func (n *netFilter) applyFiltering() error {
+	if n.config == nil {
+		return nil // Net Filtering disabled
+	}
+
+	vethChainName := vethChainPrefix + n.ifaceName
+
+	logrus.Debugf("NetFilter. Allowing ingress: %s %s for %s", n.config.allowedNets, n.config.allowedRanges, n.ifaceName)
+
+	// Verify expected chains "CONTAINERS" and "CONTAINER-REJECT" exist
+	for _, chainName := range []string{containersChainName, containerRejectChainName} {
+		if !chainExists(chainName) {
+			return fmt.Errorf("Expected iptables chain not found: %s", chainName)
+		}
+	}
+
+	rules := new(iptablesRules)
+	rules.addRule("-N", vethChainName) // create veth specific chain
+
+	// Allow specified nets and ranges only
+	for _, ipNet := range n.config.allowedNets {
+		rules.addRule("-A", vethChainName, "-s", ipNet.String(), "-j", "ACCEPT")
+	}
+	for _, ipRange := range n.config.allowedRanges {
+		rules.addRule("-A", vethChainName, "-m", "iprange", "--src-range", ipRange.String(), "-j", "ACCEPT")
+	}
+
+	rules.addRule("-A", vethChainName, "-j", "CONTAINER-REJECT")
+
+	// Add JUMP in CONTAINERS, send all traffic going to the veth interface
+	rules.addRule("-I", containersChainName, "1", "-o", n.ifaceName, "-j", vethChainName)
+
+	if err := rules.apply(); err != nil {
```

Comment:
Should the case where a rule with the same name already exist ? Could happen if a machine with containers reboot and get the same container.